File: ios/ShortDrama/Sources/Data/DTOs/AuthDTOs.swift
import Foundation

struct AuthEnvelopeDTO<T: Decodable>: Decodable, Equatable where T: Equatable {
    let code: Int
    let data: T
    let message: String
}

struct SendOtpRequestDTO: Encodable, Equatable {
    let countryCode: String
    let phone: String
    let scene: String
}

struct SendOtpResponseDTO: Decodable, Equatable {
    let requestId: String
    let cooldownSeconds: Int
    let expiresInSeconds: Int

    private enum CodingKeys: String, CodingKey {
        case requestId
        case cooldownSeconds
        case expiresInSeconds
    }

    init(requestId: String, cooldownSeconds: Int, expiresInSeconds: Int) {
        self.requestId = requestId
        self.cooldownSeconds = cooldownSeconds
        self.expiresInSeconds = expiresInSeconds
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        requestId = try container.decode(String.self, forKey: .requestId)
        cooldownSeconds = try container.decode(Int.self, forKey: .cooldownSeconds)
        expiresInSeconds = (try? container.decode(Int.self, forKey: .expiresInSeconds)) ?? 300
    }

    func toEntity() -> SendOtpResult {
        SendOtpResult(
            requestId: requestId,
            cooldownSeconds: cooldownSeconds,
            expiresInSeconds: expiresInSeconds
        )
    }
}

struct CreateAuthSessionRequestDTO: Encodable, Equatable {
    let countryCode: String
    let phone: String
    let code: String
}

struct RefreshAuthSessionRequestDTO: Encodable, Equatable {
    let refreshToken: String
}

struct AuthUserDTO: Codable, Equatable {
    let id: String
    let phone: String
    let displayName: String?
    let avatarURL: String?
    let role: String
    let isNewUser: Bool

    private enum CodingKeys: String, CodingKey {
        case id
        case phone
        case displayName
        case avatarURL = "avatarUrl"
        case role
        case isNewUser
    }

    func toEntity() -> AuthUser {
        AuthUser(
            id: id,
            phone: phone,
            displayName: displayName,
            avatarURL: avatarURL,
            role: role,
            isNewUser: isNewUser
        )
    }
}

struct AuthSessionDTO: Codable, Equatable {
    let accessToken: String
    let refreshToken: String
    let expiresAt: String
    let user: AuthUserDTO

    private enum CodingKeys: String, CodingKey {
        case accessToken
        case refreshToken
        case expiresAt
        case user
    }

    func toEntity() -> AuthSession {
        AuthSession(
            accessToken: accessToken,
            refreshToken: refreshToken,
            expiresAt: expiresAt,
            user: user.toEntity()
        )
    }
}

typealias SendOtpEnvelopeDTO = AuthEnvelopeDTO<SendOtpResponseDTO>
typealias AuthSessionEnvelopeDTO = AuthEnvelopeDTO<AuthSessionDTO>
typealias AuthUserEnvelopeDTO = AuthEnvelopeDTO<AuthUserDTO>
struct EmptySuccessDTO: Decodable, Equatable {}

File: android/app/src/main/java/com/djs66256/short_drama/data/dto/AuthDtos.kt
package com.djs66256.short_drama.data.dto

import com.djs66256.short_drama.domain.model.AuthRole
import com.djs66256.short_drama.domain.model.AuthSession
import com.djs66256.short_drama.domain.model.AuthUser
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class SendOtpRequestDto(
    @SerialName("countryCode") val countryCode: String,
    @SerialName("phone") val phone: String,
    @SerialName("scene") val scene: String,
)

@Serializable
data class CreateAuthSessionRequestDto(
    @SerialName("countryCode") val countryCode: String,
    @SerialName("phone") val phone: String,
    @SerialName("code") val code: String,
)

@Serializable
data class RefreshAuthSessionRequestDto(
    @SerialName("refreshToken") val refreshToken: String,
)

@Serializable
data class ApiEnvelopeDto<T>(
    @SerialName("code") val code: Int,
    @SerialName("data") val data: T,
    @SerialName("message") val message: String,
)

@Serializable
data class SendOtpPayloadDto(
    @SerialName("requestId") val requestId: String,
    @SerialName("cooldownSeconds") val cooldownSeconds: Int,
    @SerialName("expiresInSeconds") val expiresInSeconds: Int,
)

@Serializable
data class AuthSessionPayloadDto(
    @SerialName("accessToken") val accessToken: String,
    @SerialName("refreshToken") val refreshToken: String,
    @SerialName("expiresAt") val expiresAt: String,
    @SerialName("user") val user: AuthUserDto,
)

@Serializable
data class AuthUserDto(
    @SerialName("id") val id: String,
    @SerialName("phone") val phone: String,
    @SerialName("displayName") val displayName: String? = null,
    @SerialName("avatarUrl") val avatarUrl: String? = null,
    @SerialName("role") val role: String,
    @SerialName("isNewUser") val isNewUser: Boolean,
)

data class SendOtpResult(
    val requestId: String,
    val cooldownSeconds: Int,
    val expiresInSeconds: Int,
)

fun SendOtpPayloadDto.toDomain(): SendOtpResult = SendOtpResult(
    requestId = requestId,
    cooldownSeconds = cooldownSeconds,
    expiresInSeconds = expiresInSeconds,
)

fun AuthSessionPayloadDto.toDomain(): AuthSession = AuthSession(
    accessToken = accessToken,
    refreshToken = refreshToken,
    expiresAtIso = expiresAt,
    user = user.toDomain(),
)

fun AuthUserDto.toDomain(): AuthUser = AuthUser(
    id = id,
    phone = phone,
    displayName = displayName,
    avatarUrl = avatarUrl,
    role = role.toAuthRole(),
    isNewUser = isNewUser,
)

private fun String.toAuthRole(): AuthRole = when (lowercase()) {
    "admin" -> AuthRole.ADMIN
    "editor" -> AuthRole.EDITOR
    else -> AuthRole.VIEWER
}
